src/coreclr/vm/interoplibinterface_java.cpp

```diff
@@ -0,0 +1,236 @@
+
+    void ReleaseGCBridgeArgumentsWorker(
+        _In_ MarkCrossReferencesArgs* args)
+    {
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+        }
+        CONTRACTL_END;
+
+        // Memory was allocated for the collections by the GC.
+        // See callers of GCToEEInterface::TriggerGCBridge().
+
+        // Free memory in each of the SCCs
+        for (size_t i = 0; i < args->ComponentCount; i++)
+        {
+            free(args->Components[i].Context);
+        }
+        free(args->Components);
+        free(args->CrossReferences);
+        free(args);
+    }
+}
+
+bool Interop::IsGCBridgeActive()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    return g_GCBridgeActive;
+}
+
+void Interop::WaitForGCBridgeFinish()
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    while (g_GCBridgeActive)
+    {
+        GCX_PREEMP();
+        g_bridgeFinished->Wait(INFINITE, false);
+        // In theory, even though we waited for bridge to finish, because we are in preemptive mode
+        // the thread could have been suspended and another GC could have happened, triggering bridge
+        // processing again. In this case we would wait again for bridge processing.
+    }
+}
+
+void Interop::TriggerClientBridgeProcessing(
+    _In_ MarkCrossReferencesArgs* args)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        PRECONDITION(GCHeapUtilities::IsGCInProgress());
+    }
+    CONTRACTL_END;
+
+    if (g_GCBridgeActive)
+    {
+        // Release the memory allocated since the GCBridge
+        // is already running and we're not passing them to it.
+        ReleaseGCBridgeArgumentsWorker(args);
+        return;
+    }
+
+    bool gcBridgeTriggered = JavaNative::TriggerClientBridgeProcessing(args);
+
+    if (!gcBridgeTriggered)
+    {
+        // Release the memory allocated since the GCBridge
+        // wasn't trigger for some reason.
+        ReleaseGCBridgeArgumentsWorker(args);
+        return;
+    }
+
+    // This runs during GC while the world is stopped, no synchronisation required
+    if (g_bridgeFinished)
+    {
+        g_bridgeFinished->Reset();
+    }
+
+    // Mark the GCBridge as active.
+    g_GCBridgeActive = true;
+}
+
+void Interop::FinishCrossReferenceProcessing(
+    _In_ MarkCrossReferencesArgs *args,
+    _In_ size_t length,
+    _In_ void* unreachableObjectHandles)
+{
+    STANDARD_VM_CONTRACT;
+    _ASSERTE(g_GCBridgeActive);
+
+    // Mark the GCBridge as inactive.
+    // This much be synchronized with the GC so switch to cooperative mode.
+    {
+        GCX_COOP();
+
+        GCHeapUtilities::GetGCHeap()->NullBridgeObjectsWeakRefs(length, unreachableObjectHandles);
+
+        IGCHandleManager* pHandleManager = GCHandleUtilities::GetGCHandleManager();
+        for (size_t i = 0; i < length; i++)
+            pHandleManager->DestroyHandleOfUnknownType(((OBJECTHANDLE*)unreachableObjectHandles)[i]);
+
+        g_GCBridgeActive = FALSE;
```

Comment:
```suggestion
        g_GCBridgeActive = false;
```
Nit

Reply:
This was supposed to be `BOOL` but seems like I mixed the code. My understanding is that `BOOL` is `int32` while `bool` could be `int8` and we use `InterlockedCompareExchange` which doesn't support types smaller than 4 bytes, therefore we would have to use `BOOL`.

Having said that, using CAS on initialize seems kind of redundant to me. I think calling `JavaMarshal.Initialize` multiple times should be undefined behavior. @jkotas @AaronRobinsonMSFT Are you fine with removing the CAS during initialize ?

Reply:
I think you are confusing this with `g_Initialized`. `g_GCBridgeActive` is `bool` and other places use lower case true/false, so this comment is just a small consistency nit.

For `g_Initialized`, I do not think you really need it. You can do InterlockedCompareExchange on `g_MarkCrossReferences` to ensure that it gets initialized only once.

Reply:
>  I think calling JavaMarshal.Initialize multiple times should be undefined behavior.

It shouldn't be undefined, it is very defined and illegal. The prototype will throw if it is already initialized. This is the same approach for ObjectiveC.